### PR TITLE
Unclear parsing of skip worktree

### PR DESCRIPTION
**Pull request by gerhardol**
## Proposed changes

Parse of 'S' was done by a git-status parser not
related to skip-worktree or assume-unchanged status.

No functionality changed.
(Well a file can be both assume-unchanged and skip-worktree simultaneously now, I do not think that is possible in Git.)

## Test methodology 

- Manual

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitCommands/Git/GitItemStatusConverter.cs [hunk 1]
```diff
@@ -2,35 +2,37 @@
...
-        public const char SkippedStatus = 'S';
+        public const char TypeChangedStatus = 'T';
         public const char UnmergedStatus = 'U';
         public const char UnmodifiedStatus_v1 = ' ';
         public const char UnmodifiedStatus_v2 = '.';
+        public const char IgnoredStatus = '!';
         public const char UntrackedStatus = '?';
 
+        // Unused char, to be used to get a default object
+        public const char UnusedCharacter = '&';
```
Comment (mstv) [inline GitCommands/Git/GitItemStatusConverter.cs:20]: ❔
```suggestion
        public const char NeutralStatus = '&';
```

**Review by mstv — COMMENTED** [review]
Some approval tests fail. (Note that Igor has a conflicting PR if they need to be adapted.)

File: GitCommands/Git/GitItemStatusConverter.cs (hunk 1 above)
Comment (gerhardol) [reply GitCommands/Git/GitItemStatusConverter.cs:20]: OK
Setting this to draft to await the approval test change

**Comment by RussKie** [comment]
> (Note that Igor has a conflicting PR if they need to be adapted.)

https://github.com/gitextensions/gitextensions/pull/10993 is merged now 🎉

File: GitCommands/Git/GitItemStatus.cs [hunk 2]
```diff
@@ -51,6 +52,16 @@ public GitItemStatus(string name)
             Name = name;
         }
 
+        /// <summary>
+        /// Get a default object for an item unchanged in the WorkTree.
+        /// </summary>
+        /// <param name="name">The file name for the item.</param>
+        /// <returns>The default GitItemStatus object.</returns>
+        public static GitItemStatus Default(string name)
```
Comment (RussKie) [inline GitCommands/Git/GitItemStatus.cs:60]:
```suggestion
        public static GitItemStatus GetDefaultStatus(string name)
```

File: UnitTests/GitCommands.Tests/Git/GetAllChangedFilesOutputParserTest.cs [hunk 3]
```diff
@@ -1,5 +1,7 @@
-﻿using GitCommands;
+﻿using System.Windows.Forms;
+using GitCommands;
 using GitCommands.Git;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
 using Newtonsoft.Json;
```
Comment (RussKie) [inline UnitTests/GitCommands.Tests/Git/GetAllChangedFilesOutputParserTest.cs:5]: 🤔 Are all of these in use?

File: UnitTests/GitCommands.Tests/Git/GetAllChangedFilesOutputParserTest.cs [hunk 4]
```diff
@@ -35,5 +37,14 @@ public async Task TestGetStatusChangedFilesFromString(string testName, string st
...
                 .UseParameters(testName);
         }
+
+        [Test]
+        public async Task GitStatusDefault()
+        {
+            GitItemStatus item = GitItemStatus.Default("filename.txt");
+
+            await Verifier.VerifyJson(JsonConvert.SerializeObject(item))
+                .UseParameters("GitStatusDefault");
```
Comment (RussKie) [inline UnitTests/GitCommands.Tests/Git/GetAllChangedFilesOutputParserTest.cs:47]: 🤔  You shouldn't need this line. It's needed if you have multiple input args, and you wish to use only a subset of those.

File: GitCommands/Git/GitItemStatus.cs (hunk 2 above)
Comment (gerhardol) [reply GitCommands/Git/GitItemStatus.cs]: ok

File: UnitTests/GitCommands.Tests/Git/GetAllChangedFilesOutputParserTest.cs (hunk 4 above)
Comment (gerhardol) [reply UnitTests/GitCommands.Tests/Git/GetAllChangedFilesOutputParserTest.cs]: removed

File: UnitTests/GitCommands.Tests/Git/GetAllChangedFilesOutputParserTest.cs (hunk 3 above)
Comment (gerhardol) [reply UnitTests/GitCommands.Tests/Git/GetAllChangedFilesOutputParserTest.cs]: System.Windows.Forms was not, the others were

**Review by mstv — APPROVED** [review]

**Review by RussKie — APPROVED** [review]

File: GitCommands/Git/GitModule.cs [hunk 5]
```diff
@@ -2774,13 +2776,18 @@ private static IReadOnlyList<GitItemStatus> GetSkipWorktreeFilesFromString(strin
             foreach (string line in lsString.LazySplit('\n', StringSplitOptions.RemoveEmptyEntries))
             {
                 char statusCharacter = line[0];
-
-                string fileName = line.SubstringAfter(' ');
-                GitItemStatus gitItemStatus = GitItemStatusConverter.FromStatusCharacter(StagedStatus.WorkTree, fileName, statusCharacter);
-                if (gitItemStatus.IsSkipWorktree)
+                const char SkippedStatus = 'S';
+                const char SkippedStatusAssumeUnchanged = 's';
+                if (statusCharacter is not SkippedStatus or SkippedStatusAssumeUnchanged)
```
Comment (berrs) [inline GitCommands/Git/GitModule.cs:2781]: Hi, sorry for commenting on a closed PR but I just noticed that this condition does not look correct.

I assume the intention is to continue if statusCharacter is not one of 'S' or 's'? If so, then the condition should be:
`statusCharacter is not SkippedStatus and not SkippedStatusAssumeUnchanged`

As it is now, the loop will continue if statusCharacter is 's'.

Comment (RussKie) [reply GitCommands/Git/GitModule.cs:2781]: Parens missing?
```suggestion
                if (statusCharacter is not (SkippedStatus or SkippedStatusAssumeUnchanged))
```

Comment (gerhardol) [reply GitCommands/Git/GitModule.cs:2781]: Actually, it is intentional. AssumeUnchanged are already added to the filelist.
There should be a comment though

Comment (berrs) [reply GitCommands/Git/GitModule.cs:2781]: Interesting, I didn't know you could use _or_ in that way.

**Pull request merged**